tests/components/tibber/conftest.py

```diff
@@ -39,8 +86,88 @@ async def mock_tibber_setup(
+
+    session_mock = MagicMock()
+    session_mock.async_ensure_token_valid = AsyncMock()
+    session_mock.token = {CONF_ACCESS_TOKEN: "test-token"}
+
+    implementation_mock = MagicMock()
+
+    data_api_client_mock = MagicMock()
+    data_api_client_mock.get_all_devices = AsyncMock(return_value={})
+    data_api_client_mock.update_devices = AsyncMock(return_value={})
 
-    with patch("tibber.Tibber", return_value=tibber_mock):
+    with (
+        patch("tibber.Tibber", return_value=tibber_mock),
+        patch(
+            "homeassistant.components.tibber.async_get_config_entry_implementation",
+            return_value=implementation_mock,
+        ),
+        patch(
+            "homeassistant.components.tibber.OAuth2Session",
+            return_value=session_mock,
+        ),
+        patch(
+            "homeassistant.components.tibber.coordinator.TibberDataAPICoordinator._async_get_client",
+            return_value=data_api_client_mock,
+        ),
+    ):
         await hass.config_entries.async_setup(config_entry.entry_id)
         await hass.async_block_till_done()
         yield tibber_mock
+
+
+@pytest.fixture
+async def setup_credentials(recorder_mock: Recorder, hass: HomeAssistant) -> None:
+    """Set up application credentials for the OAuth flow."""
+    assert await async_setup_component(hass, "application_credentials", {})
+    await async_import_client_credential(
+        hass,
+        DOMAIN,
+        ClientCredential("test-client-id", "test-client-secret"),
+        DOMAIN,
+    )
+
+
+@pytest.fixture
+def data_api_entry(hass: HomeAssistant) -> MockConfigEntry:
+    """Create a Data API Tibber config entry."""
+    entry = MockConfigEntry(
+        domain=DOMAIN,
+        data={CONF_ACCESS_TOKEN: "token"},
+        unique_id="data-api",
+    )
+    entry.add_to_hass(hass)
+    return entry
+
+
+def create_mock_runtime(
+    async_get_client: AsyncMock | None = None,
+    tibber_connection: MagicMock | None = None,
+    coordinator_data: dict | None = None,
+) -> TibberRuntimeData:
+    """Create a mock TibberRuntimeData.
+
+    Args:
+        async_get_client: Optional async mock for getting the Data API client.
+        tibber_connection: Optional mock for the GraphQL connection.
+        coordinator_data: Optional data dict for the coordinator.
+
+    """
+    session = MagicMock()
+    session.async_ensure_token_valid = AsyncMock()
+    session.token = {CONF_ACCESS_TOKEN: "test-token"}
+
+    coordinator = MagicMock(spec=TibberDataAPICoordinator)
+    coordinator.data = coordinator_data if coordinator_data is not None else {}
+    coordinator.sensors_by_device = {}
+
+    runtime = MagicMock(spec=TibberRuntimeData)
+    runtime.session = session
+    runtime.tibber_connection = tibber_connection or MagicMock()
+    runtime.tibber_connection.get_homes = MagicMock(return_value=[])
+    runtime.data_api_coordinator = coordinator
+    runtime.async_get_client = async_get_client or AsyncMock()
+
+    return runtime
```

Comment:
We shouldn't touch internals, HA should setup the integration and we should patch the library, so then we can assert the state machine and other sources if it went like we expected